colossalai/pipeline/rpc/_pipeline_base.py

```diff
@@ -344,16 +332,10 @@ def subscribe_producer(self, microbatch_id: int, forward_only: bool):
                 key = UniqueKey(microbatch_id, Phase.FORWARD)
                 if key in self.work_list:
                     return
-
-            producer_stage_ids = []
-            with self.producer_consumer_init_lock:
```

Comment:
Why we need this lock previously, and why we could remove it safely now?

Reply:
Same as the last comments. In past, we calculate both of them in every worker thread and save it for possible rpc users(even in different rpc threads). Since stage relationship is fixed if and only if topo is given, I remove the def-use depends of  producer_stage_ids and consumer_stage_ids by re-calculating from topo anytime we need it.